internal/controllers/machineset/machineset_preflight.go

```diff
@@ -0,0 +1,220 @@
+	// or a control plane where the version is not set. In both cases we cannot perform any preflight checks as
+	// we do not have enough information. Return early.
+	cpVersion, err := contract.ControlPlane().Version().Get(controlPlane)
+	if err != nil {
+		if errors.Is(err, contract.ErrFieldNotFound) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q: failed to perform preflight checks: failed to get the version of ControlPlane %s", action, cpKlogRef)
+	}
+	cpSemver, err := semver.ParseTolerant(*cpVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q: failed to perform preflight checks: failed to parse version %q of ControlPlane %s", action, *cpVersion, cpKlogRef)
+	}
+
+	errList := []error{}
+	preflightCheckErrs := []preflightCheckError{}
+	// Run the control-plane-stable preflight check.
+	if !skipped.Has(clusterv1.MachineSetPreflightCheckControlPlaneStable) {
+		preflightCheckErr, err := r.controlPlaneStablePreflightCheck(controlPlane)
+		if err != nil {
+			errList = append(errList, err)
+		}
+		if preflightCheckErr != nil {
+			preflightCheckErrs = append(preflightCheckErrs, preflightCheckErr)
+		}
+	}
+
+	// Check the version skew policies only if version is defined in the MachineSet.
+	if ms.Spec.Template.Spec.Version != nil {
+		msVersion := *ms.Spec.Template.Spec.Version
+		msSemver, err := semver.ParseTolerant(msVersion)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q: failed to perform preflight checks: failed to parse version %q of MachineSet %s", action, msVersion, klog.KObj(ms))
+		}
+
+		// Run the kubernetes-version skew preflight check.
+		if !skipped.Has(clusterv1.MachineSetPreflightCheckKubernetesVersion) {
+			preflightCheckErr := r.kubernetesVersionPreflightCheck(cpSemver, msSemver)
+			if preflightCheckErr != nil {
+				preflightCheckErrs = append(preflightCheckErrs, preflightCheckErr)
+			}
+		}
+
+		// Run the kubeadm-version skew preflight check.
+		if !skipped.Has(clusterv1.MachineSetPreflightCheckKubeadmVersion) {
+			preflightCheckErr, err := r.kubeadmVersionPreflightCheck(cpSemver, msSemver, ms)
+			if err != nil {
+				errList = append(errList, err)
+			}
+			if preflightCheckErr != nil {
+				preflightCheckErrs = append(preflightCheckErrs, preflightCheckErr)
+			}
+		}
+	}
+
+	if len(errList) > 0 {
+		return ctrl.Result{}, errors.Wrapf(kerrors.NewAggregate(errList), "failed to perform %q: failed to perform preflight checks", action)
+	}
+	if len(preflightCheckErrs) > 0 {
+		preflightCheckErrStrings := []string{}
+		for _, v := range preflightCheckErrs {
+			preflightCheckErrStrings = append(preflightCheckErrStrings, *v)
+		}
+		log.Info(fmt.Sprintf("While performing %q preflight checks failed: %s", action, strings.Join(preflightCheckErrStrings, ",")))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
```

Comment:
cc @sbueringer 
given the work on removing remaining usage of requeue After

Reply:
Thx. I already thought about it and would prefer to use a requeue for now. I think this is one of the more complicated cases where it's hard to figure out if we actually get an event for all cases where the preflight checks could have failed.

E.g. is every MachineSet re-reconciled as soon as the Control Plane becomes stable.

I think let's use the requeue for now and if scale tests show that this is a problem let's iterate.